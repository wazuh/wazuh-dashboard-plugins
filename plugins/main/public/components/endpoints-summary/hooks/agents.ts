import { useState, useEffect } from 'react';
import { getTotalAgentsService } from '../services';

export const useGetTotalAgents = () => {
  const [totalAgents, setTotalAgents] = useState<number>();
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState();

  const getTotalAgents = async () => {
    try {
      setIsLoading(true);
      const totalAgents = await getTotalAgentsService();
      setTotalAgents(totalAgents);
      setError(undefined);
    } catch (error: any) {
      setError(error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    getTotalAgents();
  }, []);

  return {
    isLoading,
    totalAgents,
    error,
  };
};
