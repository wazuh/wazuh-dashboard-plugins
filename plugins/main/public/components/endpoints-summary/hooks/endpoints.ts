import { useState, useEffect } from 'react';
import { WzRequest } from '../../../react-services/wz-request';

export const useGetTotalEndpoints = () => {
  const [totalEndpoints, setTotalEndpoints] = useState<number>();
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState();

  const getTotalEndpoints = async () => {
    try {
      setIsLoading(true);
      const {
        data: {
          data: { total_affected_items },
        },
      } = await WzRequest.apiReq('GET', '/agents', {
        params: { limit: 1, q: 'id!=000' },
      });
      setTotalEndpoints(total_affected_items);
      setError(undefined);
    } catch (error: any) {
      setError(error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    getTotalEndpoints();
  }, []);

  return {
    isLoading,
    totalEndpoints,
    error,
  };
};
