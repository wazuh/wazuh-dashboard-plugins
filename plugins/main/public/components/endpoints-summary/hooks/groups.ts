import { useState, useEffect } from 'react';
import { getGroupsService } from '../services';

export const useGetGroups = () => {
  const [groups, setGroups] = useState<string[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState();

  const getGroups = async () => {
    try {
      setIsLoading(true);
      const { affected_items } = await getGroupsService();
      const groups = affected_items.map(item => item.name);
      setGroups(groups);
      setError(undefined);
    } catch (error: any) {
      setError(error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    getGroups();
  }, []);

  return {
    isLoading,
    groups,
    error,
  };
};
