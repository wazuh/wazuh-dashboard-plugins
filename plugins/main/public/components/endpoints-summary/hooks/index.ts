export { useGetTotalAgents } from './agents';
export { useGetGroups } from './groups';
