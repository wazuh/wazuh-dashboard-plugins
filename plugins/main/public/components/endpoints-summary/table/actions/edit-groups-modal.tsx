import React, { useState, useEffect } from 'react';
import {
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiButtonEmpty,
  EuiButton,
  EuiForm,
  EuiFormRow,
  EuiComboBox,
  EuiBadge,
} from '@elastic/eui';
import { compose } from 'redux';
import { withErrorBoundary, withReduxProvider } from '../../../common/hocs';
import { UI_LOGGER_LEVELS } from '../../../../../common/constants';
import { UI_ERROR_SEVERITIES } from '../../../../react-services/error-orchestrator/types';
import { getErrorOrchestrator } from '../../../../react-services/common-services';
import { useGetGroups } from '../../hooks';
import {
  addAgentToGroupService,
  removeAgentFromGroupsService,
} from '../../services';
import { Agent } from '../../types';
import { getToasts } from '../../../../kibana-services';

interface EditAgentGroupsModalProps {
  agent: Agent;
  onClose: () => void;
  reloadAgents: () => void;
}

export const EditAgentGroupsModal = compose(
  withErrorBoundary,
  withReduxProvider,
)(({ agent, onClose, reloadAgents }: EditAgentGroupsModalProps) => {
  const [selectedGroups, setSelectedGroups] = useState(
    agent.group?.map(group => ({ label: group })) || [],
  );
  const [isSaving, setIsSaving] = useState(false);

  const {
    groups,
    isLoading: isGroupsLoading,
    error: errorGroups,
  } = useGetGroups();

  if (errorGroups) {
    const options = {
      context: `EditAgentGroupsModal.useGetGroups`,
      level: UI_LOGGER_LEVELS.ERROR,
      severity: UI_ERROR_SEVERITIES.BUSINESS,
      store: true,
      error: {
        error: errorGroups,
        message: errorGroups.message || errorGroups,
        title: `Could not get groups`,
      },
    };
    getErrorOrchestrator().handleError(options);
  }

  const showToast = (
    color: string,
    title: string = '',
    text: string = '',
    time: number = 3000,
  ) => {
    getToasts().add({
      color: color,
      title: title,
      text: text,
      toastLifeTimeMs: time,
    });
  };

  const handleOnSave = async () => {
    setIsSaving(true);

    const flatSelectedGroups = selectedGroups.map(group => group.label);
    const addedGroups =
      flatSelectedGroups.filter(group => !agent.group?.includes(group)) || [];
    const removedGroups =
      agent.group?.filter(group => !flatSelectedGroups.includes(group)) || [];

    if (!removedGroups.length && !addedGroups.length) {
      setIsSaving(false);
      onClose();
      return;
    }

    try {
      addedGroups.length &&
        (await Promise.all(
          addedGroups?.map(groupId =>
            addAgentToGroupService({ agentId: agent.id, groupId }),
          ),
        ));
      removedGroups.length &&
        (await removeAgentFromGroupsService({
          agentId: agent.id,
          groupIds: removedGroups,
        }));
      showToast('success', 'Edit agent groups', 'Groups saved successfully');
      reloadAgents();
    } catch (error) {
      const options = {
        context: `EditAgentGroupsModal.handleOnSave`,
        level: UI_LOGGER_LEVELS.ERROR,
        severity: UI_ERROR_SEVERITIES.BUSINESS,
        store: true,
        error: {
          error,
          message: error.message || error,
          title: `Could not save agent groups`,
        },
      };
      getErrorOrchestrator().handleError(options);
    } finally {
      setIsSaving(false);
      onClose();
    }
  };

  const form = (
    <EuiForm component='form'>
      <EuiFormRow label='Agent'>
        <EuiBadge color='hollow'>{agent.name}</EuiBadge>
      </EuiFormRow>
      <EuiFormRow
        label='Groups'
        isRequired
        isInvalid={!selectedGroups?.length}
        error={['You must add at least one group']}
      >
        <EuiComboBox
          placeholder='Select groups'
          options={groups?.map(group => ({ label: group })) || []}
          selectedOptions={selectedGroups}
          onChange={selectedGroups => setSelectedGroups(selectedGroups)}
          isLoading={isGroupsLoading}
          clearOnBlur
        />
      </EuiFormRow>
    </EuiForm>
  );

  return (
    <EuiModal
      onClose={onClose}
      onClick={ev => {
        ev.stopPropagation();
      }}
    >
      <EuiModalHeader>
        <EuiModalHeaderTitle>Edit agent groups</EuiModalHeaderTitle>
      </EuiModalHeader>

      <EuiModalBody>{form}</EuiModalBody>

      <EuiModalFooter>
        <EuiButtonEmpty onClick={onClose}>Cancel</EuiButtonEmpty>
        <EuiButton
          onClick={handleOnSave}
          fill
          isLoading={isSaving}
          disabled={isGroupsLoading || !selectedGroups?.length}
        >
          Save
        </EuiButton>
      </EuiModalFooter>
    </EuiModal>
  );
});
