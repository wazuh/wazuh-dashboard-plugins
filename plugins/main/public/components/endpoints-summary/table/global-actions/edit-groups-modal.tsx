import React, { useState } from 'react';
import {
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiButtonEmpty,
  EuiButton,
  EuiForm,
  EuiFormRow,
  EuiComboBox,
  EuiText,
} from '@elastic/eui';
import { compose } from 'redux';
import { withErrorBoundary, withReduxProvider } from '../../../common/hocs';
import { UI_LOGGER_LEVELS } from '../../../../../common/constants';
import { UI_ERROR_SEVERITIES } from '../../../../react-services/error-orchestrator/types';
import { getErrorOrchestrator } from '../../../../react-services/common-services';
import { useGetGroups } from '../../hooks';
import {
  addAgentToGroupService,
  removeAgentFromGroupService,
} from '../../services';
import { Agent } from '../../types';
import {
  AgentsGlobalActionsResultModal,
  RESULT_TYPE,
  Result,
} from './result-modal';

interface EditAgentsGroupsModalProps {
  agents: Agent[];
  onClose: () => void;
  reloadAgents: () => void;
  addOrRemove: 'add' | 'remove';
}

type Option = {
  label: string;
};

export const EditAgentsGroupsModal = compose(
  withErrorBoundary,
  withReduxProvider,
)(
  ({
    agents,
    onClose,
    reloadAgents,
    addOrRemove,
  }: EditAgentsGroupsModalProps) => {
    const [selectedGroups, setSelectedGroups] = useState<Option[]>([]);
    const [isSaving, setIsSaving] = useState(false);
    const [isResultModalVisible, setIsResultModalVisible] = useState(false);
    const [results, setResults] = useState<Result[]>([]);

    const {
      groups,
      isLoading: isGroupsLoading,
      error: errorGroups,
    } = useGetGroups();

    if (errorGroups) {
      const options = {
        context: `EditAgentsGroupsModal.useGetGroups`,
        level: UI_LOGGER_LEVELS.ERROR,
        severity: UI_ERROR_SEVERITIES.BUSINESS,
        store: true,
        error: {
          error: errorGroups,
          message: errorGroups.message || errorGroups,
          title: `Could not get groups`,
        },
      };
      getErrorOrchestrator().handleError(options);
    }

    const getStringArrayGroups = (options: Option[]) => {
      return options.map(group => group.label);
    };

    const handleOnChange = (selectedGroups: Option[]) => {
      setSelectedGroups(selectedGroups);
    };

    const handleOnSave = async () => {
      setIsSaving(true);

      try {
        const promises = agents.reduce((acc, agent) => {
          const groups = getStringArrayGroups(selectedGroups);

          if (addOrRemove === 'add') {
            const groupsToAdd = groups.filter(
              group => !agent.group?.includes(group),
            );

            const groupsToAddPromises = groupsToAdd?.map(group =>
              addAgentToGroupService(agent.id, group)
                .then(() =>
                  setResults(results => {
                    const newResult = {
                      type: RESULT_TYPE.SUCCESS,
                      agent,
                      group,
                    };
                    return [...results, newResult];
                  }),
                )
                .catch(error => {
                  setResults(results => {
                    const newResult = {
                      type: RESULT_TYPE.ERROR,
                      agent,
                      group,
                      message: error.message,
                    };
                    return [...results, newResult];
                  });
                }),
            );

            return [...acc, ...groupsToAddPromises];
          }

          const groupsToRemove = groups.filter(group =>
            agent.group?.includes(group),
          );

          const groupsToRemovePromises = groupsToRemove?.map(group =>
            removeAgentFromGroupService(agent.id, group)
              .then(() =>
                setResults(results => {
                  const newResult = {
                    type: RESULT_TYPE.SUCCESS,
                    agent,
                    group,
                  };
                  return [...results, newResult];
                }),
              )
              .catch(error => {
                setResults(results => {
                  const newResult = {
                    type: RESULT_TYPE.ERROR,
                    agent,
                    group,
                    message: error.message,
                  };
                  return [...results, newResult];
                });
              }),
          );

          return [...acc, ...groupsToRemovePromises];
        }, [] as any);

        await Promise.allSettled(promises);

        setIsSaving(false);
        setIsResultModalVisible(true);

        reloadAgents();
      } catch (error) {
        setIsSaving(false);
        const options = {
          context: `EditAgentsGroupsModal.handleOnSave`,
          level: UI_LOGGER_LEVELS.ERROR,
          severity: UI_ERROR_SEVERITIES.BUSINESS,
          store: true,
          error: {
            error,
            message: error.message || error,
            title: `Could not save agents groups`,
          },
        };
        getErrorOrchestrator().handleError(options);
      }
    };

    const groupsText =
      addOrRemove === 'add'
        ? 'Select groups to add'
        : 'Select groups to remove';

    const form = (
      <EuiForm component='form'>
        <EuiFormRow label='Seleted agents'>
          <EuiText>{agents.length}</EuiText>
        </EuiFormRow>
        <EuiFormRow label={groupsText}>
          <EuiComboBox
            placeholder={groupsText}
            options={groups?.map(group => ({ label: group })) || []}
            selectedOptions={selectedGroups}
            onChange={handleOnChange}
            isLoading={isGroupsLoading}
            clearOnBlur
          />
        </EuiFormRow>
      </EuiForm>
    );

    return (
      <EuiModal onClose={onClose}>
        {isResultModalVisible ? (
          <AgentsGlobalActionsResultModal
            title={
              addOrRemove === 'add'
                ? 'Add groups to agents'
                : 'Remove groups from agents'
            }
            isLoading={isSaving}
            results={results}
            onClose={onClose}
            loadingMessage='Adding groups to agents...'
          />
        ) : (
          <>
            <EuiModalHeader>
              <EuiModalHeaderTitle>
                {addOrRemove === 'add'
                  ? 'Add groups to agents'
                  : 'Remove groups from agents'}
              </EuiModalHeaderTitle>
            </EuiModalHeader>
            <EuiModalBody>{form}</EuiModalBody>
            <EuiModalFooter>
              <EuiButtonEmpty onClick={onClose}>Cancel</EuiButtonEmpty>
              <EuiButton
                onClick={handleOnSave}
                fill
                isLoading={isSaving}
                disabled={isGroupsLoading || !selectedGroups?.length}
              >
                Save
              </EuiButton>
            </EuiModalFooter>
          </>
        )}
      </EuiModal>
    );
  },
);
