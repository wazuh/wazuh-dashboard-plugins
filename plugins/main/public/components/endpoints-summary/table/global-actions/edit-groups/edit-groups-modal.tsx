import React, { useState } from 'react';
import {
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiButtonEmpty,
  EuiButton,
  EuiForm,
  EuiFormRow,
  EuiComboBox,
  EuiText,
  EuiCallOut,
} from '@elastic/eui';
import { compose } from 'redux';
import { withErrorBoundary, withReduxProvider } from '../../../../common/hocs';
import { UI_LOGGER_LEVELS } from '../../../../../../common/constants';
import { UI_ERROR_SEVERITIES } from '../../../../../react-services/error-orchestrator/types';
import { getErrorOrchestrator } from '../../../../../react-services/common-services';
import { useGetGroups } from '../../../hooks';
import {
  addAgentsToGroupService,
  getAgentsService,
  removeAgentsFromGroupService,
} from '../../../services';
import { Agent } from '../../../types';
import { EditAgentsGroupsModalResult } from './result';
import { ErrorAgent } from '../../../services/paginated-agents-group';

export enum RESULT_TYPE {
  SUCCESS = 'success',
  ERROR = 'error',
}

export type GroupResult = {
  group: string;
  result: RESULT_TYPE;
  successAgents?: string[];
  errorMessage?: string;
  totalErrorAgents?: number;
  errorAgents?: ErrorAgent[];
};

interface EditAgentsGroupsModalProps {
  selectedAgents: Agent[];
  allAgentsSelected: boolean;
  filters: any;
  onClose: () => void;
  reloadAgents: () => void;
  addOrRemove: 'add' | 'remove';
}

type Option = {
  label: string;
};

export const EditAgentsGroupsModal = compose(
  withErrorBoundary,
  withReduxProvider,
)(
  ({
    selectedAgents,
    allAgentsSelected,
    filters,
    onClose,
    reloadAgents,
    addOrRemove,
  }: EditAgentsGroupsModalProps) => {
    const [selectedGroups, setSelectedGroups] = useState<Option[]>([]);
    const [finalAgents, setFinalAgents] = useState<Agent[]>([]);
    const [getAgentsStatus, setGetAgentsStatus] = useState('disabled');
    const [getAgentsError, setGetAgentsError] = useState();
    const [saveChangesStatus, setSaveChangesStatus] = useState('disabled');
    const [isResultVisible, setIsResultVisible] = useState(false);
    const [groupResults, setGroupResults] = useState<GroupResult[]>([]);

    const {
      groups,
      isLoading: isGroupsLoading,
      error: errorGetGroups,
    } = useGetGroups();

    if (errorGetGroups) {
      const options = {
        context: `EditAgentsGroupsModal.useGetGroups`,
        level: UI_LOGGER_LEVELS.ERROR,
        severity: UI_ERROR_SEVERITIES.BUSINESS,
        store: true,
        error: {
          error: errorGetGroups,
          message: errorGetGroups.message || errorGetGroups,
          title: `Could not get groups`,
        },
      };
      getErrorOrchestrator().handleError(options);
    }

    const getAgents = async () => {
      if (!allAgentsSelected) {
        setGetAgentsStatus('complete');
        return selectedAgents;
      }
      try {
        const { affected_items } = await getAgentsService({ filters });
        setGetAgentsStatus('complete');
        return affected_items;
      } catch (error) {
        setGetAgentsStatus('danger');
        setGetAgentsError(error);

        const options = {
          context: `EditAgentsGroupsModal.handleOnSave`,
          level: UI_LOGGER_LEVELS.ERROR,
          severity: UI_ERROR_SEVERITIES.BUSINESS,
          store: true,
          error: {
            error,
            message: error.message || error,
            title: `Could not get agents data`,
          },
        };
        getErrorOrchestrator().handleError(options);
      }
    };

    const getArrayByProperty = (array: any[], propertyName: string) => {
      return array.map(element => element[propertyName]);
    };

    const handleOnSave = async () => {
      setGetAgentsStatus('loading');
      setIsResultVisible(true);

      const agents = await getAgents();

      if (!agents?.length) {
        return;
      }

      setFinalAgents(agents);

      setSaveChangesStatus('loading');

      const agentIds = getArrayByProperty(agents, 'id');

      const groups = getArrayByProperty(selectedGroups, 'label');

      const promises = groups.map(group => {
        const promise =
          addOrRemove === 'add'
            ? addAgentsToGroupService({ agentIds, groupId: group })
            : removeAgentsFromGroupService({ agentIds, groupId: group });
        return promise
          .then(result => {
            const { data, error, message } = result.data;
            const { affected_items, failed_items, total_failed_items } = data;
            setGroupResults(results => {
              const newGroupResult = {
                group,
                result: error ? RESULT_TYPE.ERROR : RESULT_TYPE.SUCCESS,
                successAgents: affected_items,
                errorAgents: failed_items,
                errorMessage: message,
                totalErrorAgents: total_failed_items,
              };
              return [...results, newGroupResult];
            });
          })
          .catch(error => {
            setGroupResults(results => {
              const newResult: GroupResult = {
                group,
                result: RESULT_TYPE.ERROR,
                errorMessage: error.message,
                errorAgents: [
                  {
                    error: { message: error.message },
                    id: agentIds,
                  },
                ],
              };
              return [...results, newResult];
            });
            const options = {
              context: `EditAgentsGroupsModal.handleOnSave`,
              level: UI_LOGGER_LEVELS.ERROR,
              severity: UI_ERROR_SEVERITIES.BUSINESS,
              store: true,
              error: {
                error,
                message: error.message || error,
                title:
                  addOrRemove === 'add'
                    ? `Could not add agents to group`
                    : `Could not remove agents from group`,
              },
            };
            getErrorOrchestrator().handleError(options);
          });
      });

      try {
        await Promise.allSettled(promises);
        setSaveChangesStatus('complete');
      } catch (error) {
        setSaveChangesStatus('danger');
        const options = {
          context: `EditAgentsGroupsModal.handleOnSave`,
          level: UI_LOGGER_LEVELS.ERROR,
          severity: UI_ERROR_SEVERITIES.BUSINESS,
          store: true,
          error: {
            error,
            message: error.message || error,
            title: `Could not save agents groups`,
          },
        };
        getErrorOrchestrator().handleError(options);
      } finally {
        reloadAgents();
      }
    };

    const groupsText =
      addOrRemove === 'add'
        ? 'Select groups to add'
        : 'Select groups to remove';

    const handleOnChangeGroupsSelect = (selectedGroups: Option[]) => {
      setSelectedGroups(selectedGroups);
    };

    const selectGroupsForm = (
      <EuiForm component='form'>
        {allAgentsSelected ? (
          <EuiFormRow>
            <EuiCallOut
              color='warning'
              iconType='alert'
              title='The changes will be applied to all agents that match the filters set in the list'
            />
          </EuiFormRow>
        ) : (
          <EuiFormRow label='Selected agents'>
            <EuiText>{selectedAgents.length}</EuiText>
          </EuiFormRow>
        )}

        <EuiFormRow label={groupsText}>
          <EuiComboBox
            placeholder={groupsText}
            options={groups?.map(group => ({ label: group })) || []}
            selectedOptions={selectedGroups}
            onChange={handleOnChangeGroupsSelect}
            isLoading={isGroupsLoading}
            clearOnBlur
          />
        </EuiFormRow>
      </EuiForm>
    );

    return (
      <EuiModal onClose={onClose}>
        <EuiModalHeader>
          <EuiModalHeaderTitle>
            {addOrRemove === 'add'
              ? 'Add groups to agents'
              : 'Remove groups from agents'}
          </EuiModalHeaderTitle>
        </EuiModalHeader>
        <EuiModalBody>
          {!isResultVisible ? (
            selectGroupsForm
          ) : (
            <EditAgentsGroupsModalResult
              finalAgents={finalAgents}
              addOrRemove={addOrRemove}
              getAgentsStatus={getAgentsStatus}
              getAgentsError={getAgentsError}
              saveChangesStatus={saveChangesStatus}
              groupResults={groupResults}
              groups={getArrayByProperty(selectedGroups, 'label')}
            />
          )}
        </EuiModalBody>
        <EuiModalFooter>
          {!isResultVisible ? (
            <>
              <EuiButtonEmpty onClick={onClose}>Cancel</EuiButtonEmpty>
              <EuiButton
                onClick={handleOnSave}
                fill
                disabled={isGroupsLoading || !selectedGroups?.length}
              >
                Save
              </EuiButton>
            </>
          ) : (
            <EuiButton
              onClick={onClose}
              fill
              disabled={
                getAgentsStatus === 'loading' || saveChangesStatus === 'loading'
              }
            >
              Close
            </EuiButton>
          )}
        </EuiModalFooter>
      </EuiModal>
    );
  },
);
