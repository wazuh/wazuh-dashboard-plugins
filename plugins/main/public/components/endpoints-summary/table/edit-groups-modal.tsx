import React, { useState, useEffect } from 'react';
import {
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiButtonEmpty,
  EuiButton,
  EuiForm,
  EuiFormRow,
  EuiComboBox,
  EuiText,
} from '@elastic/eui';
import { compose } from 'redux';
import { withErrorBoundary, withReduxProvider } from '../../common/hocs';
import { UI_LOGGER_LEVELS } from '../../../../common/constants';
import { UI_ERROR_SEVERITIES } from '../../../react-services/error-orchestrator/types';
import { getErrorOrchestrator } from '../../../react-services/common-services';
import { useGetGroups } from '../hooks/groups';
import {
  addAgentToGroupService,
  removeAgentFromGroupsService,
} from '../services';

interface EditAgentGroupsModalProps {
  agent: { id: string; name: string; group: string[] };
  onClose: () => void;
  reloadAgents: () => void;
}

export const EditAgentGroupsModal = compose(
  withErrorBoundary,
  withReduxProvider,
)(({ agent, onClose, reloadAgents }: EditAgentGroupsModalProps) => {
  const [selectedGroups, setSelectedGroups] = useState(
    agent.group?.map(group => ({ label: group })) || [],
  );
  const [isSaving, setIsSaving] = useState(false);

  const {
    groups,
    isLoading: isGroupsLoading,
    error: errorGroups,
  } = useGetGroups();

  if (errorGroups) {
    const options = {
      context: `EditAgentGroupsModal.useGetGroups`,
      level: UI_LOGGER_LEVELS.ERROR,
      severity: UI_ERROR_SEVERITIES.BUSINESS,
      store: true,
      error: {
        error,
        message: error.message || error,
        title: `Could not get groups`,
      },
    };
    getErrorOrchestrator().handleError(options);
  }

  const handleOnSave = async () => {
    setIsSaving(true);

    const flatSelectedGroups = selectedGroups.map(group => group.label);
    const addedGroups =
      flatSelectedGroups.filter(group => !agent.group?.includes(group)) || [];
    const removedGroups =
      agent.group?.filter(group => !flatSelectedGroups.includes(group)) || [];

    if (!removedGroups.length && !addedGroups.length) {
      setIsSaving(false);
      onClose();
      return;
    }

    try {
      addedGroups.length &&
        (await Promise.all(
          addedGroups?.map(group => addAgentToGroupService(agent.id, group)),
        ));
      removedGroups.length &&
        (await removeAgentFromGroupsService(agent.id, removedGroups));
      reloadAgents();
    } catch (error) {
      const options = {
        context: `EditAgentGroupsModal.handleOnSave`,
        level: UI_LOGGER_LEVELS.ERROR,
        severity: UI_ERROR_SEVERITIES.BUSINESS,
        store: true,
        error: {
          error,
          message: error.message || error,
          title: `Could not save agent groups`,
        },
      };
      getErrorOrchestrator().handleError(options);
    } finally {
      setIsSaving(false);
      onClose();
    }
  };

  const form = (
    <EuiForm component='form'>
      <EuiFormRow label='Agent'>
        <EuiText>
          <b>{agent.name}</b>
        </EuiText>
      </EuiFormRow>
      <EuiFormRow
        label='Groups'
        isRequired
        isInvalid={!selectedGroups?.length}
        error={['You must add at least one group']}
      >
        <EuiComboBox
          placeholder='Select groups'
          options={groups?.map(group => ({ label: group })) || []}
          selectedOptions={selectedGroups}
          onChange={selectedGroups => setSelectedGroups(selectedGroups)}
          isLoading={isGroupsLoading}
        />
      </EuiFormRow>
    </EuiForm>
  );

  return (
    <EuiModal
      onClose={onClose}
      initialFocus='[name=popswitch]'
      onClick={ev => {
        ev.stopPropagation();
      }}
    >
      <EuiModalHeader>
        <EuiModalHeaderTitle>Edit agent groups</EuiModalHeaderTitle>
      </EuiModalHeader>

      <EuiModalBody>{form}</EuiModalBody>

      <EuiModalFooter>
        <EuiButtonEmpty onClick={onClose}>Cancel</EuiButtonEmpty>
        <EuiButton
          onClick={handleOnSave}
          fill
          isLoading={isSaving}
          disabled={isGroupsLoading || !selectedGroups?.length}
        >
          Save
        </EuiButton>
      </EuiModalFooter>
    </EuiModal>
  );
});
