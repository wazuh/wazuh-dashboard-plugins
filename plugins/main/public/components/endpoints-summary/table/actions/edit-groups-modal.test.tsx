import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { EditAgentGroupsModal } from './edit-groups-modal';
import { useGetGroups } from '../../hooks';

jest.mock('../../services', () => ({
  addAgentToGroupService: jest.fn(),
  removeAgentFromGroupsService: jest.fn(),
}));

jest.mock('../../hooks', () => ({
  useGetGroups: jest.fn(),
}));

jest.mock('../../../../react-services/common-services', () => ({
  getErrorOrchestrator: () => ({
    handleError: () => {},
  }),
}));

describe('EditAgentGroupsModal component', () => {
  test('should return the component with save disabled', async () => {
    (useGetGroups as jest.Mock).mockReturnValue({
      isLoading: true,
    });

    const { container, getByText, getByRole } = render(
      <EditAgentGroupsModal
        agent={{
          id: '001',
          name: 'agent1',
          group: ['default'],
        }}
        onClose={() => {}}
        reloadAgents={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();

    const agentName = getByText('agent1');
    expect(agentName).toBeInTheDocument();

    const saveButton = getByRole('button', { name: 'Save' });
    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toBeDisabled();

    const cancelButton = getByRole('button', { name: 'Cancel' });
    expect(cancelButton).toBeInTheDocument();
  });

  test('should return the component', async () => {
    (useGetGroups as jest.Mock).mockReturnValue({
      isLoading: false,
      groups: ['default', 'group1', 'group2'],
    });

    const { container, getByText, getByRole } = render(
      <EditAgentGroupsModal
        agent={{
          id: '001',
          name: 'agent1',
          group: ['default'],
        }}
        onClose={() => {}}
        reloadAgents={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();

    const agentName = getByText('agent1');
    expect(agentName).toBeInTheDocument();

    const saveButton = getByRole('button', { name: 'Save' });
    expect(saveButton).toBeInTheDocument();

    const cancelButton = getByRole('button', { name: 'Cancel' });
    expect(cancelButton).toBeInTheDocument();
  });

  test('should select a new group', async () => {
    (useGetGroups as jest.Mock).mockReturnValue({
      isLoading: false,
      groups: ['default', 'group1', 'group2'],
    });

    const { getByText, getByRole } = render(
      <EditAgentGroupsModal
        agent={{
          id: '001',
          name: 'agent1',
          group: ['default'],
        }}
        onClose={() => {}}
        reloadAgents={() => {}}
      />,
    );

    const saveButton = getByRole('button', { name: 'Save' });
    expect(saveButton).toBeInTheDocument();

    const selectedGroup = getByText('default');
    expect(selectedGroup).toBeInTheDocument();

    act(() => {
      fireEvent.click(selectedGroup);
    });

    await waitFor(() => expect(getByText('group1')).toBeInTheDocument());

    act(() => {
      fireEvent.click(getByText('group1'));
      fireEvent.click(saveButton);
    });
  });
});
