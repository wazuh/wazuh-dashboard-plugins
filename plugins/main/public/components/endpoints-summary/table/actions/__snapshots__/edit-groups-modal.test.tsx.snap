// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditAgentGroupsModal component should return the component 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;

exports[`EditAgentGroupsModal component should return the component with save disabled 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
